etcdctl/command/backup_noconf_command.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/codegangsta/cli"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/idutil"
+	"github.com/coreos/etcd/pkg/pbutil"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/snap"
+	"github.com/coreos/etcd/store"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+)
+
+func NewBackupNoConfCommand() cli.Command {
+	return cli.Command{
+		Name:  "backup-noconf",
+		Usage: "backup an etcd directory",
+		Flags: []cli.Flag{
+			cli.StringFlag{Name: "data-dir", Value: "", Usage: "Path to the etcd data dir"},
+			cli.StringFlag{Name: "backup-dir", Value: "", Usage: "Path to the backup dir"},
+		},
+		Action: handleBackupNoConf,
+	}
+}
+
+// handleBackup handles a request that intends to do a backup.
+func handleBackupNoConf(c *cli.Context) {
+	srcSnap := path.Join(c.String("data-dir"), "snap")
+	destSnap := path.Join(c.String("backup-dir"), "snap")
+	srcWAL := path.Join(c.String("data-dir"), "wal")
+	destWAL := path.Join(c.String("backup-dir"), "wal")
+
+	if err := os.MkdirAll(destSnap, 0700); err != nil {
+		log.Fatalf("failed creating backup snapshot dir %v: %v", destSnap, err)
+	}
+	ss := snap.New(srcSnap)
+	snapshot, err := ss.Load()
+	if err != nil && err != snap.ErrNoSnapshot {
+		log.Fatal(err)
+	}
+	var walsnap walpb.Snapshot
+	if snapshot != nil {
+		walsnap.Index, walsnap.Term = snapshot.Metadata.Index, snapshot.Metadata.Term
+		newss := snap.New(destSnap)
+		snapshot, err := purgeConfInSnap(snapshot)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if err := newss.SaveSnap(*snapshot); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	w, err := wal.OpenNotInUse(srcWAL, walsnap)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer w.Close()
+	wmetadata, state, ents, err := w.ReadAll()
+	switch err {
+	case nil:
+	case wal.ErrSnapshotNotFound:
+		fmt.Printf("Failed to find the match snapshot record %+v in wal %v.", walsnap, srcWAL)
+		fmt.Printf("etcdctl will add it back. Start auto fixing...")
+	default:
+		log.Fatal(err)
+	}
+	var metadata etcdserverpb.Metadata
+	pbutil.MustUnmarshal(&metadata, wmetadata)
+	idgen := idutil.NewGenerator(uint8(time.Now().UnixNano()), time.Now())
+	metadata.ClusterID = idgen.Next()
+	metadata.NodeID = 0
+
+	neww, err := wal.Create(destWAL, pbutil.MustMarshal(&metadata))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer neww.Close()
+	if err := neww.Save(state, purgeConfInEnts(ents)); err != nil {
+		log.Fatal(err)
+	}
+	if err := neww.SaveSnapshot(walsnap); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func purgeConfInSnap(snap *raftpb.Snapshot) (*raftpb.Snapshot, error) {
+	st := store.New()
+	err := st.Recovery(snap.Data)
+	if err != nil {
+		return nil, err
+	}
+	st.Delete("/0", true, true)
+	d, err := st.Save()
+	if err != nil {
+		return nil, err
+	}
+	snap.Data = d
+	snap.Metadata.ConfState = raftpb.ConfState{Nodes: make([]uint64, 0)}
+	return snap, nil
+}
+
+func purgeConfInEnts(ents []raftpb.Entry) []raftpb.Entry {
```

Comment:
this is doing the right thing, but could it make sense to make the entries that we skip into Noops instead of removing them?